src/sentry/search/snuba/backend.py

```diff
@@ -0,0 +1,352 @@
+            reverse=True,
+            **paginator_options
+        ).get_result(limit, cursor, count_hits=count_hits)
+
+        groups = Group.objects.in_bulk(paginator_results.results)
+        paginator_results.results = [groups[k] for k in paginator_results.results if k in groups]
+
+        return paginator_results
+
+
+def do_search(project_id, environment_id, tags, start, end,
+              sort, candidates=None, limit=1000, **parameters):
+    from sentry.search.base import ANY
+
+    filters = {
+        'project_id': [project_id],
+    }
+
+    if environment_id is not None:
+        filters['environment'] = [environment_id]
+
+    if candidates is not None:
+        hashes = list(
+            GroupHash.objects.filter(
+                group_id__in=candidates
+            ).values_list(
+                'hash', flat=True
+            ).distinct()
+        )
+
+        if not hashes:
+            return {}
+
+        filters['primary_hash'] = hashes
+
+    having = SnubaConditionBuilder({
+        'age_from': ScalarCondition('first_seen', '>'),
+        'age_to': ScalarCondition('first_seen', '<'),
+        'last_seen_from': ScalarCondition('last_seen', '>'),
+        'last_seen_to': ScalarCondition('last_seen', '<'),
+        'times_seen': CallbackCondition(
+            lambda times_seen: ('times_seen', '=', times_seen),
+        ),
+        'times_seen_lower': ScalarCondition('times_seen', '>'),
+        'times_seen_upper': ScalarCondition('times_seen', '<'),
+    }).build(parameters)
+
+    conditions = []
+    for tag, val in six.iteritems(tags):
+        col = 'tags[{}]'.format(tag)
+        if val == ANY:
+            conditions.append((col, '!=', ''))
+        else:
+            conditions.append((col, '=', val))
+
+    aggregations = [
+        ['count()', '', 'times_seen'],
+        ['min', 'timestamp', 'first_seen'],
+        ['max', 'timestamp', 'last_seen'],
+        [priority_expr, '', 'priority']
+    ]
+
+    # {hash -> {times_seen -> int
+    #           first_seen -> date_str,
+    #           last_seen -> date_str,
+    #           priority -> int},
+    #  ...}
+    snuba_results = snuba.query(
+        start=start,
+        end=end,
+        groupby=['primary_hash'],
+        conditions=conditions,
+        having=having,
+        filter_keys=filters,
+        aggregations=aggregations,
+        orderby=sort,
+        limit=limit,
+    )
+
+    # {hash -> group_id, ...}
+    hash_to_group = dict(
+        GroupHash.objects.filter(
+            project_id=project_id,
+            hash__in=snuba_results.keys()
+        ).values_list(
+            'hash', 'group_id'
+        ).distinct()
+    )
+
+    # {group_id -> {field1: [...all values from field1 for all hashes...],
+    #               field2: [...all values from field2 for all hashes...]
+    #               ...}
+    #  ...}
+    group_data = {}
+    for hash, obj in snuba_results.items():
+        if hash in hash_to_group:
```

Comment:
Hmm, do you think we should log a warning, blow up, or what our other options?

Reply:
I'm honestly not sure. I think blowing up is not the right answer since there are perfectly legitimate cases where this can happen since the two systems aren't expected to be consistent together… logging a warning at the end if we've returned hashes that didn't match to a group (and including the specific hashes so we can dig in later if need be) is probably good for now so we at least have some degree of visibility into how often it's occurring. 

Eventually we might need to trigger some sort of repair/compaction/cleanup/whatever for the project data if this number gets too high?